Mathlib/RingTheory/MvPowerSeries/Evaluation.lean

```diff
@@ -0,0 +1,360 @@
+* `MvPowerSeries.HasEval` : the `Prop`-valued structure
+that is required to evaluate power series
+
+* `MvPowerSeries.uniformContinuous_eval₂` : uniform continuity of the evaluation
+
+* `MvPowerSeries.continuous_eval₂` : continuity of the evaluation
+
+* `MvPowerSeries.aeval` : the evaluation map as an algebra map
+
+-/
+
+namespace MvPowerSeries
+
+open Topology
+
+open Filter MvPolynomial RingHom Set TopologicalSpace UniformSpace
+
+/- ## Necessary conditions -/
+
+section
+
+variable {σ : Type*}
+variable {R : Type*} [CommRing R] [TopologicalSpace R]
+variable {S : Type*} [CommRing S] [TopologicalSpace S]
+variable {φ : R →+* S}
+
+-- We endow MvPowerSeries σ R with the Pi topology
+open WithPiTopology
+
+/-- Families at which power series can be evaluated -/
+structure HasEval (a : σ → S) : Prop where
+  hpow : ∀ s, IsTopologicallyNilpotent (a s)
+  tendsto_zero : Tendsto a cofinite (𝓝 0)
+
+/-- The domain of evaluation of `MvPowerSeries`, as an ideal -/
+def hasEvalIdeal [IsTopologicalRing S] [IsLinearTopology S S] : Ideal (σ → S) where
+  carrier := {a | HasEval a}
+  add_mem' {a} {b} ha hb := {
+    hpow := fun s ↦ IsTopologicallyNilpotent.add (ha.hpow s) (hb.hpow s)
+    tendsto_zero := by
+      rw [← add_zero 0]
+      exact ha.tendsto_zero.add hb.tendsto_zero }
+  zero_mem' := {
+    hpow := fun s ↦ by
+      simp only [Pi.zero_apply]
+      apply tendsto_atTop_of_eventually_const (i₀ := 1)
+      intro i hi
+      rw [zero_pow (Nat.ne_zero_iff_zero_lt.mpr hi)]
+    tendsto_zero := tendsto_const_nhds }
+  smul_mem' c {x} hx := {
+    hpow := fun s ↦ by
+      simp only [IsTopologicallyNilpotent, Pi.smul_apply', smul_eq_mul, mul_pow]
+      exact IsLinearTopology.tendsto_mul_zero_of_right _ _ (hx.hpow s)
+    tendsto_zero := IsLinearTopology.tendsto_mul_zero_of_right _ _ hx.tendsto_zero }
+
+theorem HasEval.add [IsTopologicalRing S] [IsLinearTopology S S]
+    {a b : σ → S} (ha : HasEval a) (hb : HasEval b) : HasEval (a + b) :=
+  hasEvalIdeal.add_mem' ha hb
+
+theorem HasEval.zero [IsTopologicalRing S] [IsLinearTopology S S] : HasEval (0 : σ → S) :=
+  hasEvalIdeal.zero_mem'
+
+theorem HasEval.smul_mem [IsTopologicalRing S] [IsLinearTopology S S]
+    (c : σ → S) {x : σ → S} (hx : HasEval x) : HasEval (c • x) :=
+  hasEvalIdeal.smul_mem' c hx
+
+theorem HasEval.comp {a : σ → R} (ha : HasEval a) {ε : R →+* S} (hε : Continuous ε) :
+    HasEval (ε ∘ a) where
+  hpow s := by
+    unfold IsTopologicallyNilpotent
+    convert (Continuous.tendsto' hε 0 0 (map_zero ε)).comp (ha.hpow s) using 2
+    simp only [Function.comp_apply, map_pow]
+  tendsto_zero := (Continuous.tendsto' hε 0 0 (map_zero ε)).comp ha.tendsto_zero
+
+/-- [Bourbaki, *Algebra*, chap. 4, §4, n°3, Prop. 4 (i) (a & b)](bourbaki1981). -/
+theorem HasEval.map (hφ : Continuous φ) {a : σ → R} (ha : HasEval a) :
+    HasEval (fun s ↦ φ (a s)) where
+  hpow := fun s ↦ IsTopologicallyNilpotent.map hφ (ha.hpow s)
+  tendsto_zero := (map_zero φ ▸ hφ.tendsto 0).comp ha.tendsto_zero
```

Comment:
I just realized, aren't these two lemmas saying exactly the same thing?

Reply:
Oops…